### DISCUSSION TITLE
Memory Leak on inference

### DISCUSSION THREAD

**Discussion by TomekPro** [mindee/doctr#1422] [Q&A] [answered]
### Bug description

Running doctr for multiple images in a loop causes massive memory leak.
![image](https://github.com/mindee/doctr/assets/49069202/d297a0f0-0bc3-4c02-8a9d-ea0bd87650b5)

### Code snippet to reproduce the bug

```
import os
import tqdm
from pathlib import Path
from doctr.io import DocumentFile
from doctr.models import ocr_predictor

model = ocr_predictor(pretrained=True)

path = Path("/path/with/jpgs")
for file in tqdm.tqdm(os.listdir(path)[0:20]):
    file_path = path / file
    doc = DocumentFile.from_images(file_path)
    result = model(doc)
```

Runned in the following way:
mprof run python test.py
mprof plot

When I modified a loop so that model was initialized as well in the loop problem was still present.

Diving into the code it seems that the problem is caused by the actual pytorch inference. For example here:
![image](https://github.com/mindee/doctr/assets/49069202/5c4472e1-bbff-4d07-8faa-67daf1b147e8)

### Error traceback

As showed in the plot above.

### Environment

Tested on empty poetry environment with just 2 packages installed:
pip install "python-doctr[torch]"
pip install memory_profiler

python 3.8.10
python-doctr             0.7.0

Ubuntu 20.04
Running on cpu

### Deep Learning backend

is_tf_available: False
is_torch_available: True

**Comment by felixdittrich92** [comment]
Hi @TomekPro :wave:,
Thanks for reporting this 👍 
It should be already fixed on the main branch (v0.8.0a) -> https://github.com/mindee/doctr/pull/1357

**Comment by TomekPro** [comment]
Hi @felixdittrich92, unfortunately this problem still occurs.

```
pip uninstall python_doctr
git clone https://github.com/mindee/doctr.git
pip install -e doctr/.
```
Then when I do `pip list | grep doctr`:
`python-doctr             0.8.0a0`

The plot still looks the same:
![image](https://github.com/mindee/doctr/assets/49069202/1554d6b1-cac3-41e6-b3f1-8c7d3df4053b)

**Comment by TomekPro** [comment]
Moving to torch 2.1 cpuonly: `pip install torch==2.1.1 torchvision==0.16.1 torchaudio==2.1.1 --index-url https://download.pytorch.org/whl/cpu` slightly helps but still leakage is clear:
![image](https://github.com/mindee/doctr/assets/49069202/6941d58b-9cdf-4a28-bf2f-81a9895d8527)

**Comment by felixdittrich92** [comment]
Mh yeah i see does this leak only exits for the CRNN models could you test it also with `vitstr_small` / `parseq` or the `master` models ? 
This would be helpful to maybe limit the bug

**Comment by TomekPro** [comment]
I just tested that the problem occurs for other recognition architectures as well, what makes me thinking that this is sth around how pytorch is used in doctr. I'm looking for a solution as well.
Parseq - just slightly better than crnn.
![image](https://github.com/mindee/doctr/assets/49069202/169e4b38-794c-4033-a047-b59642d84743)

vitstr_small - the same:
![image](https://github.com/mindee/doctr/assets/49069202/a28906f4-3dd9-4521-8a30-c4d1e62df237)

master - the same:
![image](https://github.com/mindee/doctr/assets/49069202/69110aab-13ce-48e1-845b-2e1030c134e1)

torch                    2.1.1+cpu

**Comment by felixdittrich92** [comment]
@TomekPro Have you tried to pass the paths as list to `doc = DocumentFile.from_images([os.path.join(root, file) for file in os.listdir(root)])`
And specify the batch size depending on your hardware:
`model = ocr_predictor(pretrained=True, det_bs=4, reco_bs=512)` for example ?

I agree after posting your plots that's still a bug (maybe on pytorch) so only an idea you could try in the meanwhile

**Comment by felixdittrich92** [comment]
And another thing you can try: https://github.com/mindee/doctr/issues/1356#issuecomment-1771020033

**Comment by TomekPro** [comment]
Yes, I tried env variables like ONEDNN_PRIMITIVE_CACHE_CAPACITY and alike without success:/
Regarding approach you suggested:
```
model = ocr_predictor(pretrained=True)
doc = DocumentFile.from_images([os.path.join(path, file) for file in os.listdir(path)[0:20]])
result = model(doc)
```
It produces really strange result, overall memory consumption is the same, it just hits max very quickly and then it is stable Maybe this could be a clue? When increasing batch size it get's even higher. Overall, I agree that this is probably pytorch thing but still it makes really hard to use doctr in real life as application would need to be restarted very often to avoid crash due to exceeding memory.
![image](https://github.com/mindee/doctr/assets/49069202/d1839cc2-f9b8-49ff-9d3c-7469b511c59a)

**Comment by felixdittrich92** [comment]
You can also disable multiprocessing which should also lower the RAM usage a bit
See point 1 : https://mindee.github.io/doctr/using_doctr/running_on_aws.html

**Comment by felixdittrich92** [comment]
But yeah i think we need to profile it more detailed again to find the real bottleneck

**Comment by TomekPro** [comment, marked as answer]
@felixdittrich92 finally, three things are needed to fix this memory leak:
1. `export DOCTR_MULTIPROCESSING_DISABLE=TRUE`
2. `export ONEDNN_PRIMITIVE_CACHE_CAPACITY=1`
3. Upgrade torch to 2.1 (in my case cpu-only version): `pip install torch==2.1.1 torchvision==0.16.1 torchaudio==2.1.1 --index-url https://download.pytorch.org/whl/cpu`
![image](https://github.com/mindee/doctr/assets/49069202/0e1a57e3-f3ee-4ef6-8c5e-90c71ae217ad)

Thanks for your help :)

**Comment by felixdittrich92** [comment]
Nice 👍 Btw. using a smaller detection model will again reduce the mem usage for example `db_mobilenet_v3_large` if it still works well enough for your use case

**Comment by felixdittrich92** [comment]
Converted to Q&A for other people who facing the same issue :)

**Comment by balochy** [comment]
same case happened to me when using DOCTR OCR for a long period of time using pytorch models, wrapping the model in the `torch.no_grad():` helped a bit to reduce the overall memory usage but there is still a significant leak both on VRAM and RAM.
```
with torch.no_grad():
    model = (ocr_predictor(
        det_arch="db_resnet50",
        reco_arch=from_hub("Felix92/doctr-torch-parseq-multilingual-v1"),
        pretrained=True,
        straighten_pages=True,
        detect_orientation=True
    ).to(device))
```
![Screenshot 2024-03-12 at 12 57 05](https://github.com/mindee/doctr/assets/49838687/5e11a382-d672-4a49-b4d7-3767614dfecb)
![Screenshot 2024-03-12 at 12 57 28](https://github.com/mindee/doctr/assets/49838687/cc34bc3c-8a52-439b-a09c-df414e9d72a2)

I will try TomekPro's soloution and report back, however the issue still persists on 0.8.0 version

**Comment by balochy** [comment]
I've used @TomekPro's method for the past week on a production environment, with around 1000-1500 OCR requests per day, the versions are specified below:

torch~=2.2.1
torchvision~=0.17.1
python-doctr[torch]~=0.8.0

the memory leak has significantly reduced, however still present and VRAM leak stays the same.

![Screenshot 2024-04-17 at 11 48 49](https://github.com/mindee/doctr/assets/49838687/5e77106e-7efa-4f4c-ad23-794f32b95c7f)
memory had around ~3 GB of leak in a span of 5 days. 
![Screenshot 2024-04-17 at 11 49 16](https://github.com/mindee/doctr/assets/49838687/4cb597b8-ac3c-40ba-a83a-b96dd608e368)
VRAM usage also went up ~ 2 GB in the same time span.

not sure if this is to be expected or not, any suggestions? the only other solutions that might work are pytorch `inference_mode` and updating doctr to 0.8.1 .
